Mathlib/Data/List/Shortlex.lean

```diff
@@ -0,0 +1,285 @@
+    simp only [true_and] at h2
+    exact h2
+  exact fun h1 => of_lex h h1
+
+theorem _root_.List.shortlex_def {s t : List α} : Shortlex r s t ↔
+    s.length < t.length ∨ s.length = t.length ∧ List.Lex r s t := by
+  constructor
+  · intro hs
+    unfold Shortlex at hs
+    generalize hp : (s.length, s) = p at hs
+    cases hs with
+    | left b₁ b₂ h =>
+      left
+      rw [Prod.mk.injEq] at hp
+      rw [← hp.1] at h
+      exact h
+    | right a h =>
+      right
+      rw [Prod.mk.injEq] at hp
+      rw [← hp.2] at h
+      exact ⟨hp.1, h⟩
+  intro hpq
+  rcases hpq with h1 | h2
+  · exact of_length_lt h1
+  exact of_lex h2.1 h2.2
+
+theorem cons_iff [IsIrrefl α r] {a : α} {s t : List α} : Shortlex r (a :: s) (a :: t) ↔
+    Shortlex r s t := by
+  simp only [shortlex_def, length_cons, add_lt_add_iff_right, add_left_inj, List.Lex.cons_iff]
+
+@[simp]
+theorem not_nil_right {s : List α} : ¬ Shortlex r s [] := by
+  intro h
+  rcases shortlex_def.mp h with h1 | h2
+  · simp only [List.length_nil, not_lt_zero'] at h1
+  exact List.Lex.not_nil_right _ _ h2.2
+
+theorem nil_left_or_eq_nil (s : List α) : Shortlex r [] s ∨ s = [] := by
+  cases s with
+  | nil => right; rfl
+  | cons head tail =>
+    left
+    apply of_length_lt
+    simp only [List.length_nil, List.length_cons, Nat.succ_eq_add_one, lt_add_iff_pos_left,
+      add_pos_iff, zero_lt_one, or_true]
+
+@[simp]
+theorem singleton_iff (a b : α) : Shortlex r [a] [b] ↔ r a b := by
+  simp only [shortlex_def, length_singleton, lt_self_iff_false, Lex.singleton_iff, true_and,
+    false_or]
+
+instance isOrderConnected [IsOrderConnected α r] [IsTrichotomous α r] :
+    IsOrderConnected (List α) (Shortlex r) where
+  conn := by
+    intro a b c ac
+    rcases shortlex_def.mp ac with h1 | h2
+    · rcases Nat.lt_or_ge a.length b.length with h3 | h4
+      · left; exact of_length_lt h3
+      rcases Nat.lt_or_ge b.length c.length with h3 | h4
+      · right; exact of_length_lt h3
+      omega
+    rcases Nat.lt_or_ge a.length b.length with h3 | h4
+    · left; exact of_length_lt h3
+    rcases Nat.lt_or_ge b.length c.length with h3 | h4
+    · right; exact of_length_lt h3
+    have hab : a.length = b.length := by omega
+    have hbc : b.length = c.length := by omega
+    rcases List.Lex.isOrderConnected.aux r a b c h2.2 with h5 | h6
+    · left
+      exact of_lex hab h5
+    right
+    exact of_lex hbc h6
+
+instance isTrichotomous [IsTrichotomous α r] : IsTrichotomous (List α) (Shortlex r) where
+  trichotomous := by
+    intro a b
+    have : a.length < b.length ∨ a.length = b.length ∨ a.length > b.length :=
+      trichotomous a.length b.length
+    rcases this with h1 | h2 | h3
+    · left; exact of_length_lt h1
+    · induction a with
+    | nil =>
+      cases b with
+      | nil => right; left; rfl
+      | cons head tail => simp only [List.length_nil, List.length_cons, Nat.succ_eq_add_one,
+        self_eq_add_left, add_eq_zero, List.length_eq_zero, one_ne_zero, and_false] at h2
```

Comment:
These indents are unusual; better as
```suggestion
    · induction a with
      | nil =>
        cases b with
        | nil => right; left; rfl
        | cons head tail =>
          simp only [List.length_nil, List.length_cons, Nat.succ_eq_add_one,
            self_eq_add_left, add_eq_zero, List.length_eq_zero, one_ne_zero, and_false] at h2
```

Reply:
oh yup, I totally missed updating that when I switched something else